web/src/main/webapp/ui/menu.ts

```diff
@@ -249,10 +325,91 @@ export class ContextMenu extends BaseMenu<MenuItem> implements IHtmlElement {
-            cell.onclick = () => { this.hide(); mi.action(); };
-        else
+        if (mi.action != null && enabled) {
+            if (cell.classList.contains("expandableMenu")) {
+              // no operation because expandableMenu only expand its submenus
+            } else {
+              cell.onclick = () => {
+                this.hide();
+                mi.action();
+              };
+            }
+        } else {
+            cell.onclick = () => this.hide();
+        }
+    }
+
+    public addExpandableItem(mi: MenuItem): number {
+        const cell = this.addItem(mi, true);
+        cell.classList.add("expandableMenu");
+        const arrow = document.createElement("span");
+        arrow.textContent = "▸";
+        arrow.classList.add("menuArrow");
+        cell.appendChild(arrow);
+        return this.cells.length - 1;
+    }
+
+    /**
+     * Inserting subMenu must be done in the end of all parent menu insert. If the
+     * submenu exceed the length of all main menu, we need to add dummyMenu
+     * */
+    public insertSubMenu(parentIndex: number, mi: MenuItem, enabled: boolean,
+            isLastSubMenu: boolean): void {
+        if (this.subMenuCells[parentIndex] === undefined)
+            this.subMenuCells[parentIndex] = emptySubMenu();
+
+        const subMenuIndex = this.subMenuCells[parentIndex].cells.length;
+        const subMenuPlacementIdx = parentIndex + subMenuIndex;
+        if (this.rows[subMenuPlacementIdx] === undefined) {
+            const trow = this.tableBody.insertRow();
+            this.rows.push(trow);
+            const cell = trow.insertCell(0);
+            cell.classList.add("dummyMenu");
+        }
+        const cell = this.rows[subMenuPlacementIdx].insertCell(1);
+        cell.id = makeId(mi.text);
+        cell.style.textAlign = "left";
+        cell.style.display = "none";
+        cell.classList.add("menuItem");
+        if (enabled) cell.classList.remove("disabled");
+        else cell.classList.add("disabled");
+        if (mi.help != null) cell.title = mi.help;
+        if (mi.text === "---") cell.innerHTML = "<hr>";
+        else cell.textContent = mi.text;
+
+        var subIndex = this.subMenuCells[parentIndex].cells.length;
+        this.subMenuCells[parentIndex].cells.push(cell);
+        cell.onmouseenter = () => this.selectSubMenu(parentIndex, subIndex);
+        cell.onmouseleave = () => this.selectSubMenu(parentIndex, -1);
+        this.setSubMenuAction(parentIndex, subMenuIndex, mi, enabled);
+
+        // Adding upperBorder: to clear the menu when the cursor hover above 1st submenu
+        if (subMenuPlacementIdx > 0 && this.subMenuCells[parentIndex].cells.length == 1) {
+            const upperBorder = this.rows[subMenuPlacementIdx - 1].insertCell(1);
+            upperBorder.classList.add("dummyMenu");
+            upperBorder.onmouseenter = () => this.clear();
+            this.subMenuCells[parentIndex].upperBorder = upperBorder;
```

Comment:
This is the approach I chose to clear the menu. It's hard to tell when is the right time to clear the submenu because the submenu is by default not selected when the cursor expands it.
![image](https://user-images.githubusercontent.com/7077197/88838156-c1855480-d19e-11ea-82a6-ce2995f26df9.png)

Reply:
test this even if you move the mouse quickly.

Reply:
1 out of 20 trials, I do catch the bug. I will fix this with a nicer approach. I am thinking of having dummyMenu on the empty spaces above and under the displayed subMenus, these dummyMenu will be transparent and will clear the menus on mouseEnter. 

You mentioned about having rectangle overlay. In my opinion, it is more complex since I need to get the 4 corners locations of the submenu which are dynamically changing.